packages/db/src/indexes/basic-index.ts

```diff
@@ -0,0 +1,439 @@
+  public readonly supportedOperations = new Set<IndexOperation>([
+    `eq`,
+    `gt`,
+    `gte`,
+    `lt`,
+    `lte`,
+    `in`,
+  ])
+
+  // Map for O(1) equality lookups: indexedValue -> Set of PKs
+  private valueMap = new Map<any, Set<TKey>>()
+  // Sorted array of unique indexed values for range queries
+  private sortedValues: Array<any> = []
+  // Set of all indexed PKs
+  private indexedKeys = new Set<TKey>()
+  // Comparator function
+  private compareFn: (a: any, b: any) => number = defaultComparator
+
+  constructor(
+    id: number,
+    expression: BasicExpression,
+    name?: string,
+    options?: any
+  ) {
+    super(id, expression, name, options)
+    this.compareFn = options?.compareFn ?? defaultComparator
+    if (options?.compareOptions) {
+      this.compareOptions = options!.compareOptions
+    }
+  }
+
+  protected initialize(_options?: BasicIndexOptions): void {}
+
+  /**
+   * Binary search to find insertion point in sorted array
+   */
+  private findInsertionIndex(value: any): number {
+    let low = 0
+    let high = this.sortedValues.length
+
+    while (low < high) {
+      const mid = (low + high) >>> 1
+      if (this.compareFn(this.sortedValues[mid], value) < 0) {
+        low = mid + 1
+      } else {
+        high = mid
+      }
+    }
+    return low
+  }
+
+  /**
+   * Adds a value to the index
+   */
+  add(key: TKey, item: any): void {
+    let indexedValue: any
+    try {
+      indexedValue = this.evaluateIndexExpression(item)
+    } catch (error) {
+      throw new Error(
+        `Failed to evaluate index expression for key ${key}: ${error}`
+      )
+    }
+
+    const normalizedValue = normalizeValue(indexedValue)
+
+    if (this.valueMap.has(normalizedValue)) {
+      // Value already exists, just add the key to the set
+      this.valueMap.get(normalizedValue)!.add(key)
+    } else {
+      // New value - add to map and insert into sorted array
+      this.valueMap.set(normalizedValue, new Set([key]))
+
+      // Insert into sorted position
+      const insertIdx = this.findInsertionIndex(normalizedValue)
+      this.sortedValues.splice(insertIdx, 0, normalizedValue)
+    }
+
+    this.indexedKeys.add(key)
+    this.updateTimestamp()
+  }
+
+  /**
+   * Removes a value from the index
+   */
+  remove(key: TKey, item: any): void {
+    let indexedValue: any
+    try {
+      indexedValue = this.evaluateIndexExpression(item)
+    } catch (error) {
+      console.warn(
+        `Failed to evaluate index expression for key ${key} during removal:`,
+        error
+      )
+      return
+    }
+
+    const normalizedValue = normalizeValue(indexedValue)
+
+    if (this.valueMap.has(normalizedValue)) {
+      const keySet = this.valueMap.get(normalizedValue)!
+      keySet.delete(key)
+
+      if (keySet.size === 0) {
+        // No more keys for this value, remove from map and sorted array
+        this.valueMap.delete(normalizedValue)
+
+        const idx = this.findInsertionIndex(normalizedValue)
```

Comment:
This findInsertionIndex + the if test below is the logic that is needed to delete a value from a sorted array. As mentioned in a previous comment `findInsertionIndex` should be replaced by the already existing `findInsertPosition`. And we should add an additional helper function `deleteInSortedArray` to that `array-utils.ts` file that does exactly this logic. And then here just use that `deleteInSortedArray` function.